### PR TITLE
bpo-40137: Move state lookups out of the critical path

### PR DESCRIPTION
**Pull request by rhettinger**
https://bugs.python.org/issue40137

### CODE REVIEW

File: Modules/_functoolsmodule.c [hunk 1]
```diff
@@ -1203,6 +1205,8 @@ lru_cache_new(PyTypeObject *type, PyObject *args, PyObject *kw)
     obj->func = func;
     obj->misses = obj->hits = 0;
     obj->maxsize = maxsize;
+    obj->kwd_mark = state->kwd_mark;                        // Borrowed
+    obj->lru_list_elem_type = state->lru_list_elem_type;    // Borrowed
```
Comment (vstinner) [inline Modules/_functoolsmodule.c]: Would it possible to use a strong reference instead? Maybe the module instance can be deleted but the object survives longer. You need to update lru_cache_tp_traverse() and lru_cache_tp_clear() in this case.

**Review by vstinner — COMMENTED** [review]
Since it's a cache, I think that adding two pointers to lru_cache_object structure is a reasonable CPU vs memory trade-off. Maybe we should reuse this approach in other performance critical code paths.

cc @corona10 @shihai1991

File: Modules/_functoolsmodule.c (hunk 1 above)
Comment (rhettinger) [reply Modules/_functoolsmodule.c]: Done.

File: Modules/_functoolsmodule.c [hunk 2]
```diff
@@ -1194,6 +1189,13 @@ lru_cache_new(PyTypeObject *type, PyObject *args, PyObject *kw)
         return NULL;
     }
 
+    state = get_functools_state_by_type(Py_TYPE(obj));
+    if (state == NULL) {
+        Py_DECREF(cachedict);
+        Py_DECREF(obj);
```
Comment (vstinner) [inline Modules/_functoolsmodule.c]: Hum. I am not sure if it's safe to call lru_cache_dealloc() here, obj is not fully initialized. For example, lru_cache_unlink_list() can crash, no? Maybe call get_functools_state_by_type(type) before creating obj?

Comment (rhettinger) [reply Modules/_functoolsmodule.c]: Done

**Review by vstinner — APPROVED** [review]
LGTM. Thanks, it looks safe with strong references.

**Pull request merged**